### PR TITLE
[EVM] Add mechanism for informing the FE of the required spill area size

### PR DESCRIPTION
**Pull request by PavelKopyl**
This update adds support for notifying the front-end (FE) about the spill area size required by the stackifier. It introduces the following components:

1. A new C API allows the FE to register a handler that the stackifier calls when a stack-too-deep (STD) error occurs

 ```
 typedef void (*LLVMStackErrorHandlerEVM)(uint64_t SpillAreaSize);
 void LLVMInstallEVMStackErrorHandler(LLVMStackErrorHandlerEVM Handler);

```
2. Function for triggering the handler. The stackifier can invoke the registered handler using:

`  void report_evm_stack_error(uint64_t spill_area_size);`

Spill/Reload Workflow.

 - FE sets up a stack error handler:

```
void on_std(uint64_t SpillAreaSize) { 
  ... // Handle bookkeeping for SpillAreaSize
  exit(1);
}
```

- FE compiles the LLVM IR normally, without specifying spill-related options
- If no STD error is encountered, compilation completes successfully and object code is generated
- If a stack-too-deep error occurs:
  - The `EVMFinalizeStackFrames` pass calls `report_evm_stack_error(size)` with the required spill area size 
  - The FE captures and stores this size, gracefully terminating the compilation thread/process(?)
  - The FE updates the memory layout and restarts the compilation process, this time supplying:
    - `StackRegionSize` - the allocated spill area size
    - `StackRegionOffset` -  the memory start address of the allocated spill area

# Code Review Checklist

## Requirements
- [ ] Have the requirements been met?
- [ ] Have stakeholder(s) approved the change?

## Implementation
- [ ] Does this code change accomplish what it is supposed to do?
- [ ] Can this solution be simplified?
- [ ] Does this change add unwanted compile-time or run-time dependencies?
- [ ] Could an additional framework, API, library, or service improve the solution?
- [ ] Could we reuse part of LLVM instead of implementing the patch or a part of it?
- [ ] Is the code at the right abstraction level?
- [ ] Is the code modular enough?
- [ ] Can a better solution be found in terms of maintainability, readability, performance, or security?
- [ ] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [ ] Are there any best practices, design patterns or language-specific patterns that could substantially improve this code? 

## Logic Errors and Bugs
- [ ] Can you think of any use case in which the
code does not behave as intended?
- [ ] Can you think of any inputs or external events
that could break the code?

## Error Handling and Logging
- [ ] Is error handling done the correct way?
- [ ] Should any logging or debugging information
be added or removed?
- [ ] Are error messages user-friendly?
- [ ] Are there enough log events and are they
written in a way that allows for easy
debugging?

## Maintainability
- [ ] Is the code easy to read?
- [ ] Is the code not repeated (DRY Principle)?
- [ ] Is the code method/class not too long?

## Dependencies
- [ ] Were updates to documentation, configuration, or readme files made as required by this change?
- [ ] Are there any potential impacts on other parts of the system or backward compatibility?

## Security
- [ ] Does the code introduce any security vulnerabilities?

## Performance
- [ ] Do you think this code change decreases
system performance?
- [ ] Do you see any potential to improve the
performance of the code significantly?

## Testing and Testability
- [ ] Is the code testable?
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
	- [ ] For changes to mutable state
- [ ] Do tests reasonably cover the code change (unit/integration/system tests)? 
	- [ ] Line Coverage
	- [ ] Region Coverage
	- [ ] Branch Coverage
- [ ] Are there some test cases, input or edge cases
that should be tested in addition?

## Readability
- [ ] Is the code easy to understand?
- [ ] Which parts were confusing to you and why?
- [ ] Can the readability of the code be improved by
smaller methods?
- [ ] Can the readability of the code be improved by
different function, method or variable names?
- [ ] Is the code located in the right
file/folder/package?
- [ ] Do you think certain methods should be
restructured to have a more intuitive control
flow?
- [ ] Is the data flow understandable?
- [ ] Are there redundant or outdated comments?
- [ ] Could some comments convey the message
better?
- [ ] Would more comments make the code more
understandable?
- [ ] Could some comments be removed by making the code itself more readable?
- [ ] Is there any commented-out code?
- [ ] Have you run a spelling and grammar checker?

## Documentation
- [ ] Is there sufficient documentation?
- [ ] Is the ReadMe.md file up to date?

## Best Practices
- [ ] Follow Single Responsibility principle?
- [ ] Are different errors handled correctly?
- [ ] Are errors and warnings logged?
- [ ] Magic values avoided?
- [ ] No unnecessary comments?
- [ ] Minimal nesting used?

## Experts' Opinion
- [ ] Do you think a specific expert, like a security
expert or a usability expert, should look over
the code before it can be accepted?
- [ ] Will this code change impact different teams, and should they review the change as well?

### CODE REVIEW

File: llvm/lib/Target/EVM/EVMFinalizeStackFrames.cpp [hunk 1]
```diff
@@ -171,10 +171,14 @@ bool EVMFinalizeStackFrames::runOnModule(Module &M) {
 
   // Check if it is valid to replace frame indices.
   if (TotalStackSize > 0) {
-    if (TotalStackSize > StackRegionSize)
+    if (TotalStackSize > StackRegionSize) {
+      // If the front-end has registered the handler, the following function
+      // should not return. Otherwise, it has no effect.
+      report_evm_stack_error(TotalStackSize);
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMFinalizeStackFrames.cpp]: It's desirable to make default `report_evm_stack_error` behavior equivalent to `report_fatal_error` below and remove that line.

**Review by akiramenai — COMMENTED** [review]
Overall, the design is much cleaner than before.
I propose that report_evm_stack_error should, by default, behave exactly like report_fatal_error, with the option for clients to override it when necessary. This keeps the code generator independent of driver/llc policy.

File: llvm/lib/Target/EVM/EVMFinalizeStackFrames.cpp (hunk 1 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMFinalizeStackFrames.cpp]: Do you mean that the `report_evm_stack_error ` should call `report_fatal_error` if no STD handler was registered:
```
void llvm::report_evm_stack_error(uint64_t spill_area_size, const Twine &Reason) {
  ...
  if (handler)
    handler(spill_area_size);
    
  // If we got here, the FE handler wasn't set up, or hasn't called abort()/exit().
  // In any case call the default error handler.
  report_fatal_error(Reason);
}
```
?

Comment (akiramenai) [reply llvm/lib/Target/EVM/EVMFinalizeStackFrames.cpp]: Yes. Something alike.

File: llvm/lib/Target/EVM/EVMFinalizeStackFrames.cpp [hunk 2]
```diff
@@ -171,10 +171,13 @@ bool EVMFinalizeStackFrames::runOnModule(Module &M) {
 
   // Check if it is valid to replace frame indices.
   if (TotalStackSize > 0) {
-    if (TotalStackSize > StackRegionSize)
-      report_fatal_error("Total stack size: " + Twine(TotalStackSize) +
-                         " is larger than the allocated stack region size: " +
-                         Twine(StackRegionSize));
+    if (TotalStackSize > StackRegionSize) {
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMFinalizeStackFrames.cpp:174]: Could we unify the terminology? We previously introduced term StackRegion and now we start using StackArea.

Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMFinalizeStackFrames.cpp:174]: Thank you, will change.

**Review by akiramenai — APPROVED** [review]
LGTM

**Review by hedgar2017 — APPROVED** [review]
